onnxruntime/__init__.py

```diff
@@ -77,79 +77,100 @@
-        # Collect all directories under site-packages/nvidia that contain .dll files (for Windows)
-        for root, _, files in os.walk(nvidia_path):
-            # Add the current directory to the DLL search path
-
-            with os.add_dll_directory(root):
-                # Find all .dll files in the current directory
-                for file in files:
-                    for pattern in lib_pattern.items().values():
-                        if pattern.match(file):
-                            dll_path = os.path.join(root, file)
-                            try:
-                                _ = ctypes.CDLL(dll_path)
-                            except Exception as e:
-                                logging.error(f"Failed to load {dll_path}: {e}")
+        # Define the list of DLL patterns, nvrtc, curand and nvJitLink are not included for Windows
+        if (11, 0) <= cuda_version() < (12, 0):
+            cuda_libs = (
+                "cublasLT64_11.dll",
+                "cublas64_11.dll",
+                "cufft64_10.dll",
+                "cudart64_11.dll",
+                "cudnn64_8.dll",
+            )
+        elif (12, 0) <= cuda_version() < (13, 0):
+            cuda_libs = (
+                "cublasLT64_12.dll",
+                "cublas64_12.dll",
+                "cufft64_11.dll",
+                "cudart64_12.dll",
+                "cudnn64_9.dll",
+            )
     elif platform.system() == "Linux":
-        # Define the patterns with optional version number and case-insensitivity
-        cuda_libs = (
-            "libcublas.so",
-            "libcublasLt.so",
-            "libcudnn.so",
-            "libcudart.so",
-            "libnvrtc.so",
-            "libcufft.so",
-            "libcurand.so",
-            "libnvJitLink.so",
-        )
-
-        # Regular expression to match .so files with optional versioning (e.g., .so, .so.1, .so.2.3)
-        lib_pattern = {pattern: re.compile(rf"{re.escape(pattern)}(\.\d+)*$", re.IGNORECASE) for pattern in cuda_libs}
-
-        # Traverse the directory and subdirectories
-        for root, _, files in os.walk(nvidia_path):
-            for file in files:
-                # Check if the file matches the .so pattern
-                for regex in lib_pattern.items().values():
-                    if regex.match(file):  # Check if the file matches the pattern
-                        so_path = os.path.join(root, file)
-                        _ = ctypes.CDLL(so_path)
+        if (11, 0) <= cuda_version() < (12, 0):
+            # Define the patterns with optional version number and case-insensitivity
+            cuda_libs = (
+                "libcublaslt.so.11",
+                "libcublas.so.11",
+                "libcurand.so.10",
+                "libcufft.so.10",
+                "libcudart.so.11",
+                "libcudnn.so.8",
+                "libnvrtc.so.11",
```

Comment:
"libnvrtc.so.11.2", #This is not a mistake, it links to more specific cuda version

Reply:
Maybe better to use regex here? In case the minor version for libnvrtc.so.11 required for cuda 11 changes?
This also applies to libnvrtc.so.12.